src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+                value < -max_int64)) then
+            error(('Input number value %q for %q field does ' ..
+                'not fit in int64'):format(value, data_spec['name']))
+        end
+        check_integer(data_spec, value)
+        return encoder.encode_I64(data_spec, value)
+    else
+        error(('Field %q of %q type gets %q type value. ' ..
+            'Unsupported or colliding types'):format(data_spec['name'],
+            data_type, type(value)))
+    end
+end
+
+-- Encodes the entered data in accordance with the
+-- selected protocol into binary format.
+--
+-- Accepts a protocol created by protobuf.protocol function,
+-- a name of a message selected for encoding and
+-- the data that needs to be encoded in the following format.
+--
+-- data = {
+--     <field_name> = <value>,
+--     <...>
+-- }
+local function encode(protocol, name, data)
+    local coded_data = {}
+    local unknown_fields = {}
+    local unknown_fields_counter = 0
+    local message_protocol = protocol[name]
+    if message_protocol == nil then
+        error(('Wrong message name %q for this protocol'):format(name))
+    end
+    local fields = message_protocol['field_by_name']
+    for field_name, value in pairs(data) do
+        if fields[field_name] == nil then
+            local data_spec = {type = 'bytes', id = unknown_fields_counter}
+            unknown_fields_counter = unknown_fields_counter + 1
+            table.insert(unknown_fields, code(data_spec, tostring(value)))
+        else
+            local field_type = fields[field_name]['type']
+            local field_num = fields[field_name]['id']
+            if protobuf_types[field_type] then
+                table.insert(coded_data, code(fields[field_name], value))
+            elseif protocol[field_type]['type'] == 'enum' then
+                local values = protocol[field_type]['value_by_name']
+                if type(value) ~= 'number' and values[value] == nil then
+                    error(('%q is not defined in %q enum'):format(value,
+                        field_type))
+                end
+                local data_spec = {type = 'int32', id = field_num}
+                if type(value) == 'number' then
+                    table.insert(coded_data, code(data_spec, value))
+                else
+                    table.insert(coded_data, code(data_spec, values[value]))
+                end
+            else
+                local coded_msg = encode(protocol, field_type, value)
+                local data_spec = {type = 'message', id = field_num}
+                table.insert(coded_data, code(data_spec, coded_msg))
+            end
+        end
+    end
+    table.insert(coded_data, table.concat(unknown_fields))
+    return table.concat(coded_data)
+end
+
+protocol_mt = {
+    __index = {
+        encode = encode,
+    }
+}
+
+return {
+    message = message,
+    enum = enum,
+    protocol = protocol,
+    encode = encode,
```

Comment:
We agreed to add the `encode` protocol object method and don't add the `encode` function. Let's drop it so.

If you need to use some internal function directly for testing purposes, let's expose it with an underscored name and hide using a `__serialize` metamethod. Examples are [here](https://github.com/tarantool/tarantool/blob/0beee2a3e2cf7b53f42c3de7a04baf22bd3cb9e7/src/box/lua/config/init.lua#L65-L85) and [here](https://github.com/tarantool/tarantool/blob/0beee2a3e2cf7b53f42c3de7a04baf22bd3cb9e7/src/lua/init.lua#L197-L213).

However, I think that it is not our case and we don't need it for testing.

Reply:
Fixed